kerncraft/iaca_marker.py

```diff
@@ -191,31 +191,124 @@ def insert_markers(asm_lines, start_line, end_line):
-    lines = list(map(str.strip, lines))
-    blocks = find_asm_blocks(lines)
+def iaca_instrumentation(input_file, output_file=None,
+                         block_selection='auto',
+                         pointer_increment='auto_with_manual_fallback'):
+    """
+    Add IACA markers to an assembly file. If instrumentation fails
+    because loop increment could not determined automatically, a ValueError is raised
+    
+    :param input_file: path to assembly file used as input
+    :param output_file: output path, if None the input is overwritten
+    :param block_selection: index of the assembly block to instrument, or 'auto' for automatically using block with the
+                            most vector instructions, or 'manual' to read index to prompt user
+    :param pointer_increment: number of bytes the pointer is incremented after the loop or 
+                              - 'auto': automatic detection, RuntimeError is raised in case of failure
+                              - 'auto_with_manual_fallback': automatic detection, fallback to manual input
+                              - 'manual': prompt user
+    :return: the instrumented assembly block
+    """
+    if output_file is None:
+        output_file = input_file
+
+    with open(input_file, 'r') as f:
+        assembly = f.readlines()
+
+    blocks = find_asm_blocks(assembly)
+    if block_selection == 'auto':
+        block_idx = select_best_block(blocks)
+    elif block_selection == 'manual':
+        block_idx = userselect_block(blocks, default=select_best_block(blocks))
+    elif isinstance(block_selection, int):
+        block_idx = block_selection
+    else:
+        raise ValueError("block_selection has to be an integer, 'auto' or 'manual' ")
 
-    # TODO check for already present markers
+    block = blocks[block_idx][1]
 
-    # Choose best default block:
-    best_idx = select_best_block(blocks)
-    # Let user select block:
-    block_idx = userselect_block(blocks, best_idx)
+    if pointer_increment == 'auto':
+        if block['pointer_increment'] is None:
+            raise RuntimeError("pointer_increment could not be detected automatically")
+    elif pointer_increment == 'auto_with_manual_fallback':
+        if block['pointer_increment'] is None:
+            block['pointer_increment'] = userselect_increment(block)
+    elif pointer_increment == 'manual':
+        block['pointer_increment'] = userselect_increment(block)
+    elif isinstance(pointer_increment, int):
+        block['pointer_increment'] = pointer_increment
+    else:
+        raise ValueError("pointer_increment has to be an integer, 'auto', 'manual' or  'auto_with_manual_fallback' ")
+
+    instrumentedAsm = insert_markers(assembly, block['first_line'], block['last_line'])
+    with open(output_file, 'w') as in_file:
+        in_file.writelines(instrumentedAsm)
+
+    return block
+
+
+def iaca_analyse_instrumented_binary(instrumented_binary_file, micro_architecture):
```

Comment:
Since this now handles also execution and parsing of IACA, I would suggest renaming it to iaca.py but still provide the iaca_marker CLI